### ISSUE TITLE
Gradient clipping is not consistent with other frameworks

### ISSUE THREAD

**Issue by pmichel31415** [clab/dynet#1374]
Right now dynet clips gradients separately for each parameters. This means that if you clip your gradients to 1.0 but have 16 `Parameter` objects, the total gradient will be clipped to a norm of 4.

The correct thing to do would be to clip the total norm, like pytorch does https://pytorch.org/docs/master/_modules/torch/nn/utils/clip_grad.html

I've found that it makes a difference (in LM experiments)

**Comment by pmichel31415** [comment]
OK after looking at the code again it looks like dynet does the same thing as pytorch, however I get better result if I clip manually with this function:

```python
def clip(pc, norm):
     tot_norm = 0.0
     for p in pc.parameters_list():
         tot_norm += (p.grad_as_array()**2).sum()
     tot_norm = np.sqrt(tot_norm)
     clip_coeff = norm / (tot_norm + 1e-6)
     if clip_coeff <1:
         for p in pc.parameters_list():
             p.scale_gradient(clip_coeff)
```

I'll leave this issue open while I investigate

**Comment by neubig** [comment]
I'd like to know what's up here, but as you noted I believe that DyNet is consistent with other frameworks.
